### PR TITLE
Add route.spec.subdomain to support implicit hostnames

### PR DESCRIPTION
**Pull request by smarterclayton**
Setting `spec.host` requires a user know the exact subdomain of
the ingress controller that will be used which is not portable
across clusters. For infrastructure components we want to have
consistent default names like `openshift-console`,
`openshift-registry`, etc which shouldn't require the operator to
introspect the default ingress controller or global config.

Instead, add a new optional field that if host is unset is
interpreted as a suggestion to the ingress controller to place
a name within its space. This field, like host, may only be
changed post creation by a user with permission since it has
the same ordering problems across namespaces. `subdomain`
can be set to `openshift-console` and the router should add
`apps.foo.com` to it (instead of following the default rule
`NAME-NAMESPACE`). Since it is a new field, its use is
optional and routers may ignore it in practice.

@ironcladlou

### CODE REVIEW

**Comment by ironcladlou** [comment]
@openshift/sig-network-edge

File: route/v1/types.go [hunk 1]
```diff
@@ -70,7 +70,23 @@ type RouteSpec struct {
 	// chosen.
 	// Must follow DNS952 subdomain conventions.
 	Host string `json:"host" protobuf:"bytes,1,opt,name=host"`
-	// Path that the router watches for, to route traffic for to the service. Optional
+	// subdomain is a DNS subdomain that is requested within the ingress controller's
+	// domain (as a subdomain). If host is set this field is ignored. An ingress
+	// controller may choose to ignore this suggested name, in which case the controller
+	// will report the assigned name in the status.ingress array or refuse to admit the
```
Comment (ironcladlou) [inline route/v1/types.go]: Looking at this and the existing docs as a whole, is it clear that regardless of whether Host or Subdomain is in use, the admitted/effective host name is reflected in `.status.ingress`?

File: route/v1/types.go [hunk 2]
```diff
@@ -70,7 +70,23 @@ type RouteSpec struct {
 	// chosen.
 	// Must follow DNS952 subdomain conventions.
 	Host string `json:"host" protobuf:"bytes,1,opt,name=host"`
-	// Path that the router watches for, to route traffic for to the service. Optional
+	// subdomain is a DNS subdomain that is requested within the ingress controller's
```
Comment (ironcladlou) [inline route/v1/types.go]: From the WIP [router implementation](https://github.com/openshift/origin/pull/22304):

> subdomain must conform to DNS 952 subdomain conventions

Should that be made explicit here?

File: route/v1/types.go (hunk 1 above)
Comment (smarterclayton) [reply route/v1/types.go]: Tes

File: route/v1/types.go (hunk 2 above)
Comment (smarterclayton) [reply route/v1/types.go]: Is dns subdomain enough?  I dunno how many users actually look up rfcs

Comment (ironcladlou) [reply route/v1/types.go]: wfm

**Comment by deads2k** [comment]
Gotcha.  This didn't make it to origin and is failing the rebase. I plan to revert and reopen

**Pull request merged**